src/base/dev/vga/vgaemu.c

```diff
@@ -1684,8 +1685,29 @@ int vga_emu_pre_init(void)
 
   vga.mem.lfb_base = 0;
   if(config.X_lfb) {
-    dosaddr_t p = alias_mapping_high(MAPPING_VGAEMU,
-				vga.mem.size, VGA_EMU_RW_PROT, vga.mem.base);
+    dosaddr_t p = 0;
+    if (config.cpu_vm_dpmi == CPUVM_KVM) {
+      /* create mapping:
+	 Linear vga.mem.lfb_base=VGAEMU_PHYS_LFB_BASE=0xe0000000 ->
+	 physical 0xe0000000 ->	addr (hugepage aligned) -> vga.mem.base,
+	 so MEM_BASE32(vga.mem.lfb_base) is not a valid pointer. */
+      p = VGAEMU_PHYS_LFB_BASE;
+      void *addr = mmap_mapping(MAPPING_VGAEMU, (void *)-1, vga.mem.size,
```

Comment:
Is this because of MAP_32BIT?
Why you need an alias for KVM?

Reply:
No, it's independent of MAP_32BIT. The problem is really that `vga.mem.base` isn't huge-page aligned, otherwise I could indeed map directly to it, instead of using the intermediate mapping.

So I could align vga.mem.base, that just needs some work in the mapping drivers?

Reply:
> No, it's independent of MAP_32BIT.

Except that it is set on line 463, and
then you seem to go to
do_mmap_huge_page_aligned()
with it still set.
And unless I am missing something,
this is not what you wanted to do.

Reply:
line 463 sets `flags = _MAP_32BIT;` but this is equal to 0 for KVM (only set for JIT and native DPMI), that's why I ignored it (and in my testing I see the map high up:
```
KVM: map slot=0 flags=0 dosaddr=0xff000000 size=0x00417000 unixaddr=0x7f0500e00000
KVM: map slot=1 flags=0 dosaddr=0x00110000 size=0x20000000 unixaddr=0x7f04daf10000
KVM: map slot=2 flags=0 dosaddr=0x00000000 size=0x000a0000 unixaddr=0x7f04dac00000
KVM: map slot=3 flags=1 dosaddr=0xe0000000 size=0x00400000 unixaddr=0x7f04d8e00000
KVM: map slot=4 flags=0 dosaddr=0x000c0000 size=0x00050000 unixaddr=0x7f04dacc0000
KVM: map slot=5 flags=1 dosaddr=0x000a0000 size=0x00020000 unixaddr=0x7f04daca0000
```
)

Reply:
OK but I propose to stop passing flags
to do_mmap_huge_page_aligned(). I
don't think it is interested in
MAP_FIXED_NOREPLACE|MAP_32BIT,
and these seems the only flags that may
be passed there.

With that it would be fine, but I of course
more prefer to get rid of an alias here.
Depends on how difficult it is to implement.

Reply:
If you would need to hack all mapping
drivers individually, then alias if perfectly
fine. If some centralized change is possible,
then good too.

Reply:
So whatever you prefer actually.
Alignment looks like a good excuse
for alias, so just get rid of the flags
please. :)